### PR TITLE
Select next best converge value compatible with our candidate set

### PR DESCRIPTION
**Pull request by Kubuxu**
Based on recommendations of Least Authority

### CODE REVIEW

File: gpbft/gpbft.go [hunk 1]
```diff
@@ -544,35 +544,37 @@ func (i *instance) tryConverge() error {
...
+		// If it is in candidate set
+		if i.isCandidate(cv.Chain) {
+			return true
+		}
+		// If it is not a candidate but it could possibly have been decided by another participant
+		// in the last round, consider it a candidate.
+		if cv.Justification.Vote.Step != PREPARE_PHASE {
+			return false
+		}
+		possibleDecisionLastRound := i.getRound(i.round-1).committed.CouldReachStrongQuorumFor(
```
Comment (Stebalien) [inline gpbft/gpbft.go:557]: Let's call `i.getRound(i.round - 1)` outside the function.

File: gpbft/gpbft.go [hunk 2]
```diff
@@ -544,35 +544,37 @@ func (i *instance) tryConverge() error {
...
+			cv.Chain.Key(), true)
+		return possibleDecisionLastRound
+	}
+
+	winner := i.getRound(i.round).converged.FindMaxTicketProposal(i.powerTable, isValidConvergeValue)
```
Comment (Stebalien) [inline gpbft/gpbft.go:562]: nit: rename to `FindBestTicketProposal`

File: gpbft/gpbft.go [hunk 3]
```diff
@@ -1332,16 +1331,17 @@ func (c *convergeState) Receive(sender ActorID, table PowerTable, value ECChain,
 
 // FindMaxTicketProposal finds the value with the best ticket, weighted by
 // sender power. Returns an invalid (zero-value) ConvergeValue if no converge is found.
```
Comment (Stebalien) [inline gpbft/gpbft.go:1333]: document filter

File: gpbft/gpbft.go [hunk 4]
```diff
@@ -1332,16 +1331,17 @@ func (c *convergeState) Receive(sender ActorID, table PowerTable, value ECChain,
...
-func (c *convergeState) FindMaxTicketProposal(table PowerTable) ConvergeValue {
+func (c *convergeState) FindMaxTicketProposal(table PowerTable, filter func(ConvergeValue) bool) ConvergeValue {
```
Comment (Stebalien) [inline gpbft/gpbft.go:1334]: I still think we should filter when receiving proposals instead of filtering here.

**Review by Stebalien — APPROVED** [review]

File: gpbft/gpbft.go (hunk 4 above)
Comment (Kubuxu) [reply gpbft/gpbft.go:1334]: I thought about it, but we should not because of how COMMIT works (the commit round is always active).

Comment (Stebalien) [reply gpbft/gpbft.go]: This is CONVERGE. We skip here: https://github.com/filecoin-project/go-f3/blob/71f7e0d13749f6f4368608640a7cd506ecf3ab3b/gpbft/gpbft.go#L358

Comment (Kubuxu) [reply gpbft/gpbft.go]: The thing that dictates sway behaviour is the state of previous commit round.

Comment (Stebalien) [reply gpbft/gpbft.go]: Kuba is right...

**Review by Stebalien — APPROVED** [review]

**Pull request merged**